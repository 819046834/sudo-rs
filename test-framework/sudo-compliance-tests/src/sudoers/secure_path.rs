use sudo_test::{Command, Env, TextFile};

use crate::{helpers, Result, SUDOERS_ALL_ALL_NOPASSWD};

#[test]
fn if_unset_searches_program_in_invoking_users_path() -> Result<()> {
    let path = "/root/my-script";
    let env = Env(SUDOERS_ALL_ALL_NOPASSWD)
        .file(path, TextFile("#!/bin/sh").chmod("100"))
        .build()?;

    Command::new("sh")
        .args(["-c", "export PATH=/root; cd /; /usr/bin/sudo my-script"])
        .exec(&env)?
        .assert_success()?;

    Ok(())
}

#[test]
fn if_set_searches_program_in_secure_path() -> Result<()> {
    let path = "/root/my-script";
    let env = Env("\
Defaults secure_path=.:/root
ALL ALL=(ALL:ALL) NOPASSWD: ALL")
    .file(path, TextFile("#!/bin/sh").chmod("100"))
    .build()?;

    let relative_path = "unset PATH; cd /bin; /usr/bin/sudo true";
    let absolute_path = "unset PATH; cd /; /usr/bin/sudo my-script";

    let scripts = [relative_path, absolute_path];

    for script in scripts {
        println!("{script}");

        Command::new("sh")
            .args(["-c", script])
            .exec(&env)?
            .assert_success()?;
    }

    Ok(())
}

#[ignore]
#[test]
fn if_set_it_becomes_the_path_set_for_program_execution() -> Result<()> {
    let secure_path = ".:/root";
    let env = Env(format!(
        "Defaults secure_path={secure_path}
ALL ALL=(ALL:ALL) NOPASSWD: ALL"
    ))
    .build()?;

    let user_path_set = "cd /; sudo /usr/bin/env";
    let user_path_unset = "unset PATH; cd /; /usr/bin/sudo /usr/bin/env";
    let scripts = [user_path_set, user_path_unset];

    for script in scripts {
        println!("{script}");

        let env_output = Command::new("sh")
            .args(["-c", script])
            .exec(&env)?
            .stdout()?;

        let env_output = helpers::parse_env_output(&env_output)?;
        assert_eq!(&secure_path, &env_output["PATH"]);
    }

    Ok(())
}
