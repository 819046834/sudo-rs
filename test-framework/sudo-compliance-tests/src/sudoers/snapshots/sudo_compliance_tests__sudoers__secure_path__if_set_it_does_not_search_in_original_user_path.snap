---
source: sudo-compliance-tests/src/sudoers/secure_path.rs
expression: output.stderr()
---
sudo: true: command not found
